### DISCUSSION TITLE
Performance issues with uplink to NPM public registry

### DISCUSSION THREAD

**Discussion by johndc7** [verdaccio/verdaccio#3043] [Ideas]
**Your Environment**
  
 * **verdaccio version**: 5.7.0
 * **node version** v16.14.0:
 * **package manager**: npm@8.3.1
 * **os**: windows 11 / linux (Tested on my Windows PC / Linux server both with clean installs)
 * **platform**: npm

**Describe the bug**
I tested this with a clean install of Verdaccio and no packages published. I have cache disabled for my uplink to the public npm registry. I was testing to make sure everything was working as expected (I just set this up today) and ran `npx create-react-app test` to see how it would handle it. I immediately noticed a huge spike in CPU usage and the download was taking a long time. I found a few issues related to performance since there's now an extra stop in between npm and the and the registry but I feel that this should not be the case when the cache is disabled.

My understanding of this is that Verdaccio is functioning as a proxy between npm and the registry. It can take longer since Verdaccio has some steps of it's own (related to caching the packages). When the cache is disabled, the extra steps should also be disabled and eliminate the need for Verdaccio to decompress and recompress the packages and the speed should be the same as using a normal proxy. I setup a basic proxy just to test this and there is no noticeable increase in CPU usage or download time.

**To Reproduce**
Clean install and install anything using an uplink to the npm public registry with cache disabled. (I used `npx create-react-app test`).

**Expected behavior**
Package should install without noticeable increase in CPU usage / download time.

**Screenshots, server logs, package manager log**
N/A: There are no errors in the logs

**Configuration File (cat ~/.config/verdaccio/config.yaml)**
I tested this using the default config. I only changed the setting to disable the cache.

**Environment information**
```
Environment Info:
  System:
    OS: Windows 10 10.0.22000
    CPU: (6) x64 Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz
  Binaries:
    npm: 8.3.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.22000.120.0), Chromium (98.0.1108.55)
```

**Comment by juanpicado** [comment]
Thanks for the report.

I'm moving this to a discussion because there are no clear steps to improve performance or where exactly could be the issue.  Feel free to debug profile that would be helpful to identify what is causing those spikes. 

More perf tickets https://github.com/verdaccio/verdaccio/issues?q=is%3Aopen+is%3Aissue+label%3Aperformance

**Comment by johndc7** [comment]
So I spent some time today digging through this and I've come to the conclusion that the performance issue comes mostly from the parsing / generating all of the JSON for the package info. When I make a request manually to Verdaccio (for example: `http://localhost:8000/verdaccio`) I can still see a little bit of a CPU spike. Since alot of packages have a pretty decent amount of info, parsing these files over and over could use a decent amount of resources.

So my current goal here (and feel free to provide any insight since I really don't have much experience how Verdaccio is designed at this point) is to get any requests for package info to proxy directly to the NPM public registry and stop Verdaccio from ever actually parsing the info.

Why? Because if we aren't actually caching any of the tarballs then why should we waste the resources on parsing the JSON and changing all of the links so they go through Verdaccio? The way I see it, if the tarballs are downloaded directly from the NPM registry then that's even less work that Verdaccio has to do anyways.

**Comment by johndc7** [comment]
So I ended up not using this since I decided that it might be helpful in the future to use more types of repositories than just npm.

What I ended up doing would also be applicable to `verdaccio` so I thought I would share.

I now have my private registry sitting behind a custom reverse proxy. For the sake of an example, lets say I'm trying to install `express`.

1. Proxy gets a request for `https://registry.example.com/express`.
2. Proxy sends a request to `http://verdaccio:8000/express`. If this request comes back with a status code higher than 401, the response is ignored and client is redirected to `https://registry.npmjs.com/express`. If the status code is less than 401, the response is sent to the client.

Since the proxy is ignoring the actual data and only looking at the status codes, there is no need to ever parse any of the data. This could be implemented into `verdaccio` for a significant speed increase when caching is disabled. You could also proxy the response from the npm public registry instead of redirecting if you wanted to do this with more than 2 registries.

Just food for thought @juanpicado

**Reply by juanpicado** [reply]
Just a quick reply to avoid ghosting your efforts, I won't have much time for this topic in the next weeks (I also will be in 🏖️ ) but I'll back to this whenever I have some space. Thanks for diving into it.
